### PR TITLE
Esversion normalization

### PR DESCRIPTION
**Pull request by jugglinmike**
This patch contains two refactoring commits followed by a bug fix that should
resolve both gh-2909 and ~~gh-3051~~ gh-2902.

### CODE REVIEW

**Comment by rwaldron** [comment]
@jugglinmike just want to make sure you've seen this

**Comment by jugglinmike** [comment]
@rwaldron You were right: somehow, I missed the build failure when I first
submitted this patch. I just had a chance to resolve it--see the `fixup!`
commit.

This is failing Coveralls because I removed a few branches that were covered
but are no longer necessary. Clicking through to the report, you'll see the
following:

Coverage | File          | Lines     | Relevant  | Covered
---------|---------------|-----------|-----------|----------
v 97.07  | src/jshint.js | 5548 (+6) | 2865 (-3) | 2781 (-3)

Working the numbers out:

    master:
    2784 / 2868 = 0.9707112970711297

    this branch:
    2781 / 2865 = 0.9706806282722513

So the "coverage" statistic does decrease, but it doesn't actually reflect any
newly-uncovered branches.

If it looks good to you, say the word and I will be happy to perform
the fixup and force-push to this branch.

**Comment by rwaldron** [comment]
> This patch contains two refactoring commits followed by a bug fix that should
resolve both gh-2909 and gh-3051.

Does that mean that this supersedes #3051?

**Comment by jugglinmike** [comment]
Sorry, @rwaldron. gh-3501 is unrelated to this work--that reference was
incorrect. The final commit in this patch should resolve both gh-2909 and
gh-2902.

File: src/jshint.js [hunk 1]
```diff
@@ -191,7 +191,10 @@ var JSHINT = (function() {
     }
   }
 
-  function assume() {
+  /**
+   * Apply all linting options according to the status of the `state` object.
+   */
+  function applyOptions() {
```
Comment (rwaldron) [inline src/jshint.js]: Much better

File: src/jshint.js [hunk 2]
```diff
@@ -417,12 +420,22 @@ var JSHINT = (function() {
...
-    var nt = state.tokens.next;
-    var body = nt.body.split(",").map(function(s) { return s.trim(); });
-
+  /**
+   * Process an inline linting directive
+   *
+   * @param {Token} dirToken - the directive-bearing comment token
+   * @param {Token} previous - the token that preceeds the directive
+   */
+  function lintingDirective(dirToken, previous) {
```
Comment (rwaldron) [inline src/jshint.js]: Also better

File: src/jshint.js [hunk 3]
```diff
@@ -452,27 +465,27 @@ var JSHINT = (function() {
...
       for (var key in predef) {
         if (_.has(predef, key)) {
-          declared[key] = nt;
+          declared[key] = dirToken;
         }
       }
     }
 
-    if (nt.type === "exported") {
+    if (dirToken.type === "exported") {
```
Comment (rwaldron) [inline src/jshint.js]: nit: I'd like to request that this be renamed: `dirToken` => `directiveToken`

**Comment by jugglinmike** [comment]
@rwaldron I've changed that variable name as requested. Sorry for the delay!

**Comment by rwaldron** [comment]
Sorry Mike, I'm the absolutely worst.

**Pull request merged**